tensorflow_federated/python/core/impl/executors/federating_executor.py

```diff
@@ -453,6 +454,353 @@ def validate_executor_placements(cls, executor_placements):
+            computation_types.NamedTupleType((fn_type, val_type))))
+
+  async def federated_value_at_server(self, arg):
+    return await self._place(arg, placement_literals.SERVER)
+
+  async def federated_value_at_clients(self, arg):
+    return await self._place(arg, placement_literals.CLIENTS)
+
+  async def federated_eval_at_server(self, arg):
+    return await self._eval(arg, placement_literals.SERVER, True)
+
+  async def federated_eval_at_clients(self, arg):
+    return await self._eval(arg, placement_literals.CLIENTS, False)
+
+  async def federated_apply(self, arg):
+    return await self._map(arg)
+
+  async def federated_map(self, arg):
+    return await self._map(arg, all_equal=False)
+
+  async def federated_map_all_equal(self, arg):
+    return await self._map(arg, all_equal=True)
+
+  async def federated_broadcast(self, arg):
+    py_typecheck.check_type(arg.type_signature, computation_types.FederatedType)
+    py_typecheck.check_type(arg.internal_representation, list)
+    if not arg.type_signature.all_equal:
+      raise ValueError('Cannot broadcast a non all_equal value.')
+    if len(arg.internal_representation) != 1:
+      raise ValueError(
+          'Cannot broadcast a with a non-singleton representation.')
+    val = await arg.internal_representation[0].compute()
+    return FederatingExecutorValue(
+        await asyncio.gather(*[
+            c.create_value(val, arg.type_signature.member)
+            for c in self._get_child_executors(placement_literals.CLIENTS)
+        ]), type_factory.at_clients(arg.type_signature.member, all_equal=True))
+
+  async def federated_zip_at_server(self, arg):
+    return await self._zip(arg, placement_literals.SERVER, all_equal=True)
+
+  async def federated_zip_at_clients(self, arg):
+    return await self._zip(arg, placement_literals.CLIENTS, all_equal=False)
+
+  async def federated_reduce(self, arg):
+    self._check_arg_is_anonymous_tuple(arg)
+    if len(arg.internal_representation) != 3:
+      raise ValueError(
+          'Expected 3 elements in the `federated_reduce()` argument tuple, '
+          'found {}.'.format(len(arg.internal_representation)))
+
+    # Encrypt client tensors
+    pk_a, sk_a = await self._trusted_aggregator_generate_keys()
+    enc_clients_arg = await self._encrypt_client_tensors(arg, pk_a)
+    val_type = enc_clients_arg.type_signature
+    val = enc_clients_arg.internal_representation
+    # Store original client dtype before encryption for
+    # future decryption
+    orig_clients_dtype = arg.type_signature[0].member.dtype
+
+    py_typecheck.check_type(val_type, computation_types.FederatedType)
+    item_type = val_type.member
+    zero_type = arg.type_signature[1]
+    op_type = arg.type_signature[2]
+
+    py_typecheck.check_type(val, list)
+    aggr = self._get_child_executors(placement_literals.AGGREGATOR, index=0)
+    server = self._get_child_executors(placement_literals.SERVER, index=0)
+
+    async def _move(v, target_executor):
```

Comment:
you will see in the merged branch that this has moved to become a classmethod on the base IntrinsicStrategy.

I would consider handling all of the encryption, zipping, & decryption in that method (and changing it from classmethod to normal method, as needed). this may be a decently-sized change, so happy to have a conversation about pros/cons, or to do this in a later PR if we decide it's a good change

Reply:
I will open an issue regarding this tomorrow.